File: findAnagrams.js
const findAnagrams = function(s, p) {
    const plen = p.length
    const result = []
    const d = {}
    let count = 0
    for (let i = 0; i < plen; i++) {
        const item = p[i]
        if (item in d) {
            d[item] += 1
        } else {
            count += 1
            d[item] = 1
        }
    }
    let result = []
    let left = 0
    for (let right = 0; right < s.length; right++) {
        if (d[s[right]] != null) d[s[right]]--
        if (d[s[right]]==0) count--
        if (count == 0) result.push(left)
        if (right - left + 1 == plen) {
            if (d[s[left]] != null) d[s[left]]++
            if (d[s[left++]] == 1) count++
        }
    }
    return result
}

File: findAnagrams.py
def findAnagrams(s, p):
    plen = len(p)
    result = []
    dicts = {}
    for char in p:
        if char in dicts:
            dicts[char] += 1
        else:
            dicts[char] = 1
    count = len(dicts)
    left = 0
    for right, char in enumerate(s):
        if char in dicts:
            dicts[char] -= 1
            if dicts[char] == 0:
                count -= 1
        if right - left + 1 == plen:
            if count == 0:
                result.append(left)
            if s[left] in dicts:
                if dicts[s[left]] == 0:
                    count += 1
                dicts[s[left]] += 1
            left += 1
    return result